### PR TITLE
Rogue stickyban detection system.

### PR DESCRIPTION
**Pull request by MrsTonedOne**
:cl:
experiment: Added a system to detect when stickybans go rogue and revert all recent matches.
/:cl:

This will detect when stickybans go rogue and revert them to whatever their state was when the world started. (if it's a legitimate rogue ban, the system has ways of shutting that all down)

I can't exactly test this locally, so before we test merge it or something, I'd like to get as many eyes on this looking for any logic errors

### CODE REVIEW

File: code/modules/admin/IsBanned.dm [hunk 1]
```diff
@@ -99,17 +105,83 @@
...
+				)
+				if (cachedban["reverting"])
+					return null
+				cachedban["reverting"] = TRUE
+
+				world.SetConfig("ban", bannedckey, null)
+
+				log_game("Stickyban on [bannedckey] detected as rogue, reverting to it's roundstart state")
+				message_admins("Stickyban on [bannedckey] detected as rogue, reverting to it's roundstart state")
+				//do not convert to timer.
```
Comment (coiax) [inline code/modules/admin/IsBanned.dm]: eyyyy

File: code/modules/admin/IsBanned.dm [hunk 2]
```diff
@@ -99,17 +105,83 @@
...
-			addclientmessage(ckey,"<span class='adminnotice'>You have been allowed to bypass a matching host/sticky ban</span>")
+			log_admin("The admin [key] has been allowed to bypass a matching host/sticky ban on [bannedckey]")
+			message_admins("<span class='adminnotice'>The admin [key] has been allowed to bypass a matching host/sticky ban on [bannedckey]</span>")
+			addclientmessage(ckey,"<span class='adminnotice'>You have been allowed to bypass a matching host/sticky ban on [bannedckey]</span>")
 			return null
-		else
-			log_access("Failed Login: [key] [computer_id] [address] - Banned [.["message"]]")
+
+		if (C) //user is already connected!.
+			C << "You are about to get disconnected for matching a sticky ban after you connected. If this turns out to be the ban evasion detection system going haywire, we will automatically detect this and revert the matches. if you feel that this is the case, please wait EXACTLY 6 seconds then reconnect using file -> reconnect to see if the match was reversed."
```
Comment (coiax) [inline code/modules/admin/IsBanned.dm]: What happens if they try to connect less than or more than 6 seconds?

Comment (MrStonedOne) [inline code/modules/admin/IsBanned.dm]: Nothing, I just want then to actually wait. If you say wait n seconds they'll just try again immediately.

**Comment by Saegrimr** [comment]
A couple things about the stickybans anyway that i've noticed.

It -always- catches the right person, and is always the next entry. THEN it adds all the false matches.

So for example, some guy gets put on a month ago.

Ckey: ShittyMcDickerson -  ban evasion. List of ckeys:
[-] Shitballs11111111 <- Him trying again but not triggering it lets say a week ago.
[-] Assmasterrrrrr2  <- This would be his newest ckey that also triggered the rogue bans right now
[-] Normaldude <- This and everybody below is a false positive
[-] Someadmin
[-] Notgoingtoreadthesticky
And so on.

If this works, could you have it keep whatever the first entry was added in the list of fuck that usually happens?

**Comment by MrStonedOne** [comment]
Think about it, if it matches the ckey at the round start state, then it will still match them after the revert.

And legitimate matches are stored in the hub anyways, so we could remove everything and just re-add the ckey with none of the other matches and it will still match them.

**Comment by Saegrimr** [comment]
That's a good point.
Didn't realize the off-server wouldn't list the match, I guess I just got lucky with picking the server and finding the first match.

**Comment by MrStonedOne** [comment]
found a few bugs, gotta fix them.

**Comment by MrStonedOne** [comment]
This ***could*** be merged in its current state, the only bug is that new stickybans aren't protected, and the only non-snowflake way to do it involves refactoring the stickyban panel and making SSstickyban process to check for new stickybans.

I've been having some minor medical issues that's limiting the time I can devote to coding.

**Comment by KorPhaeron** [comment]
@optimumtact 

Is this good to merge

**Comment by optimumtact** [comment]
Yes

**Pull request merged**